be/src/exprs/agg/group_concat.h

```diff
@@ -292,4 +299,398 @@ class GroupConcatAggregateFunction
+            array_nulls[j] = &(nullable_arrays[j]->null_column_data());
+            array_nulls[j]->resize(old_size + chunk_size);
+            array_offsets[j] = &(arrays[j]->offsets_column()->get_data());
+        }
+        for (auto i = 0; i < chunk_size; i++) {
+            if (null_data[i]) {
+                for (auto j = 0; j < columns.size(); ++j) {
+                    (*array_nulls[j])[i + old_size] = 1;
+                    array_offsets[j]->push_back(array_offsets[j]->back());
+                }
+            } else {
+                for (auto j = 0; j < columns.size(); ++j) {
+                    (*array_nulls[j])[i + old_size] = 0;
+                    arrays[j]->elements_column()->append_datum(src[j]->get(i));
+                    array_offsets[j]->push_back(array_offsets[j]->back() + 1);
+                }
+            }
+        }
+        for (auto j = 0; j < columns.size(); ++j) {
+            nullable_arrays[j]->update_has_null();
+        }
+    }
+
+    // group_concat(a, b order by c, d), output a,b,',' by c and d, but ignore the last separator ','
+    // empty state return null, other output row by row.
+    // note as output columns and order-by columns are put in group-by clause if specify DISTINCT, so here need to do
+    // distinct further after order by data columns.
+    void finalize_to_column(FunctionContext* ctx, ConstAggDataPtr __restrict state, Column* to) const override {
+        auto& state_impl = this->data(state);
+        if (state_impl.data_columns == nullptr) {
+            to->append_default();
+            return;
+        }
+        auto elem_size = (*state_impl.data_columns)[0]->size();
+        if (elem_size == 0) {
+            to->append_default();
+            return;
+        }
+        auto output_col_num = state_impl.output_col_num + 1; // include sep
+        Columns outputs(output_col_num);
+        for (auto i = 0; i < output_col_num; ++i) {
+            outputs[i] = (*state_impl.data_columns)[i];
+        }
+        // order by
+        if (!ctx->get_is_asc_order().empty()) {
+            for (auto i = 0; i < output_col_num; ++i) {
+                outputs[i] = (*state_impl.data_columns)[i]->clone_empty();
+            }
+            Permutation perm;
+            Columns order_by_columns;
+            SortDescs sort_desc(ctx->get_is_asc_order(), ctx->get_nulls_first());
+            order_by_columns.assign(state_impl.data_columns->begin() + output_col_num, state_impl.data_columns->end());
+            Status st = sort_and_tie_columns(ctx->state()->cancelled_ref(), order_by_columns, sort_desc, &perm);
+            // release order-by columns early
+            order_by_columns.clear();
+            state_impl.release_order_by_columns();
+            DCHECK(ctx->state()->cancelled_ref() || st.ok());
+            for (auto i = 0; i < output_col_num; ++i) {
+                materialize_column_by_permutation(outputs[i].get(), {(*state_impl.data_columns)[i]}, perm);
```

Comment:
is that possible  late materialize column in the final output?

Reply:
it is determited by the repeated ratio for a chunk. if more repeated tuples, do distinct first is better, otherwise sort first is better.